### PR TITLE
[FIX][website_sale] Avoid compute tax for every company

### PR DESCRIPTION
**Pull request by chienandalu**
This patch filters the company set on the company so taxes are computed just for that company.

Description of the issue/feature this PR addresses:

When a product has multiple company taxes and b2c prices is enabled, price is not computed properly and adds to the base price every company tax present. 

Current behavior before PR:

For instance, if I had in a product two company taxes the calculation is something like:

`pricelist_price + tax_company_1 + tax_company_2`

Desired behavior after PR is merged:

Only taxes belonging to the active website are computed.

cc @Tecnativa

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_sale/models/product.py [hunk 1]
```diff
@@ -191,14 +191,15 @@ def _website_price(self):
         qty = self._context.get('quantity', 1.0)
         partner = self.env.user.partner_id
         pricelist = self.env['website'].get_current_website().get_current_pricelist()
+        company_id = self.env['website'].get_current_website().company_id
```
Comment (pedrobaeza) [inline addons/website_sale/models/product.py]: As you are using `get_current_website` 2 times, store its value in an intermediate variable.

File: addons/website_sale/models/product.py [hunk 2]
```diff
@@ -191,14 +191,15 @@ def _website_price(self):
...
 
         context = dict(self._context, pricelist=pricelist.id, partner=partner)
         self2 = self.with_context(context) if self._context != context else self
 
         ret = self.env.user.has_group('sale.group_show_price_subtotal') and 'total_excluded' or 'total_included'
 
         for p, p2 in zip(self, self2):
-            taxes = partner.property_account_position_id.map_tax(p.taxes_id)
+            taxes = partner.property_account_position_id.map_tax(p.taxes_id).filtered(lambda x: x.company_id == company_id)
```
Comment (pedrobaeza) [inline addons/website_sale/models/product.py]: Put the filter before the call to `map_tax`

**Comment by pedrobaeza** [comment]
@JKE-be can you please check this one?

**Comment by pedrobaeza** [comment]
I don't know the problem in CLA, but he is included here: https://github.com/odoo/odoo/blob/10.0/doc/cla/corporate/tecnativa.md

File: addons/website_sale/models/product.py (hunk 2 above)
Comment (online-odoo) [inline addons/website_sale/models/product.py]: @pedrobaeza 
what do you mean by 'put filter before'?
It is map_tax that returns the list of taxes, before the map tax, you just have one 'account.fiscal.position'

Comment (JKE-be) [inline addons/website_sale/models/product.py]: @pedrobaeza
what do you mean by 'put filter before'?
It is map_tax that returns the list of taxes, before the map tax, you just have one 'account.fiscal.position'

Comment (pedrobaeza) [inline addons/website_sale/models/product.py]: I mean:
```
taxes = partner.property_account_position_id.map_tax(p.taxes_id.filtered(lambda x: x.company_id == company_id))
```
This way, you save some computations mapping invalid taxes, unless cross-companies fiscal position tax mappings is something allowed...

Comment (JKE-be) [inline addons/website_sale/models/product.py]: :+1:

File: addons/website_sale/models/product.py [hunk 3]
```diff
@@ -200,7 +200,7 @@ def _website_price(self):
         ret = self.env.user.has_group('sale.group_show_price_subtotal') and 'total_excluded' or 'total_included'
 
         for p, p2 in zip(self, self2):
-            taxes = partner.property_account_position_id.map_tax(p.taxes_id).filtered(lambda x: x.company_id == company_id)
+            taxes = partner.property_account_position_id.map_tax(p.sudo().taxes_id.filtered(lambda x: x.company_id == company_id))
```
Comment (pedrobaeza) [inline addons/website_sale/models/product.py]: Why `sudo()`? Please don't mix multi-company issues. That should be treated in another issue/PR. Any way, I don't see here the problem without sudo, unless the user of `p` is not admin nor public user.

Comment (chienandalu) [inline addons/website_sale/models/product.py]: @pedrobaeza The problem I'm finding here is not with multi company buy with permissions. In the case of a default `public user` session: `product` is loaded with `public user` `uid` and can't run methods from taxes object due to permissions. Instead, partner is loaded with `admin` `uid`, so when `filtered()`  is piped on it permissions aren't a problem.

As you pointed before this issue roots are deeper than what this patch solves so maybe the solution must be refocused.

Comment (pedrobaeza) [inline addons/website_sale/models/product.py]: OK, but let's try to refocus first this PR. So are you saying that without multi-company, you're having permission issues in this code? If so, then the solution is to add the corresponding ACLs/record rules for accessing the involved objects. Can you put the steps to reproduce the problem in runbot?

Comment (chienandalu) [inline addons/website_sale/models/product.py]: @pedrobaeza I'm commiting the patch without `sudo()` so you can better check behaviour. Also summarizing the behavior this are the steps to reproduce (company1 and company2 for the example):

- If no `sudo()` is applied:

1.  403 error trying to access to the shop due to permission of public user.

- Before patch:

1. Enable b2c prices
2. Create tax1 linked to company1
3. Create tax2 linked to company2
4. From company1 add tax1 to productA
5. From company2 add tax2 to productA
6. If a user is present in both companies (either public or not): price = list_price + tax1 + tax2

- After patch:

1. Same first steps.
2. If a user is present in both companies (either public or not): price = list_price + tax(belonging to website company)

Comment (pedrobaeza) [inline addons/website_sale/models/product.py]: But the problem `403 error trying to access to the shop due to permission of public user.` is reproducible with only 1 company?

Comment (chienandalu) [inline addons/website_sale/models/product.py]: @pedrobaeza Yes, the problem still shows up with one company.

Steps:

1. Enable b2c prices
2. Add customer tax to product.
3. With public user, error 500 (i put 403 before wich wasn't right, sorry) log:
```
AccessError: (u'Sorry, you are not allowed to access this document. Only users with the following access level are currently allowed to do that:\n- Employees/Employee\n\t- Accounting & Finance/Billing\n\t- Accounting & Finance/Adviser\n\t- Sales/User: Own Documents Only\n\t- Sales/User: Own Documents Only\n\t- Other Extra Rights/Portal\n\n(Document model: account.tax)', None)
```
4. No issues being other type of user.

So I'll research now in another PR the minimum ACL needed for public user to give public user such permissions and leave this one pending until that one is ackwoledge.

Comment (pedrobaeza) [inline addons/website_sale/models/product.py]: OK, please don't forget to include these STR on that PR.

**Comment by JKE-be** [comment]
@pedrobaeza 
  CLA fail because commited with gmail address instead of david.vidal@tecnativa.com 

> From e2d3ad8bdd377a2ceb6a6c428bc847823567332e Mon Sep 17 00:00:00 2001
> From: David Vidal <chienandalu@gmail.com>
> Date: Mon, 14 Aug 2017 12:08:48 +0200

**Comment by pedrobaeza** [comment]
OK, that's right. Thanks for checking.

@chienandalu please change your commit address. You can do it with `git commit --amend --author "David Vidal <david.vidal@tecnativa.com>"`

**Comment by JKE-be** [comment]
Closed by https://github.com/odoo/odoo/commit/04cddc447b89906ed3001c89993a90e934b05adf

typo in commit msg

**Comment by pedrobaeza** [comment]
Thank you very much!

**Comment by chienandalu** [comment]
Thanks @JKE-be! Keep also in mind this related patch :slightly_smiling_face: : https://github.com/odoo/odoo/pull/18881

**Comment by KangOl** [comment]
I revert commit 04cddc447b89906ed3001c89993a90e934b05adf as public user can't access /shop/ anymore

**Comment by pedrobaeza** [comment]
Uhm, it's because #18881 is not still merged.

**Comment by KangOl** [comment]
We can't merge it as it require an update of the database to be effective.

**Comment by pedrobaeza** [comment]
But the problem is that current state makes is a bomb waiting for exploding in any moment with a patch. The real problem is that public user doesn't have access to taxes, which is a requirement when dealing with B2C/B2B prices. It's a coincidence that previous sentence doesn't provoke also the access error.

Any way, maybe we can bypass the stability restriction including here a `sudo()` for just that specific line. What do you think?

**Comment by KangOl** [comment]
The use of `sudo()` is a solution but need to be carefully tested if it doesn't introduce other company missmatch issues (when user's company is not the same as admin's)

**Comment by pedrobaeza** [comment]
Yeah, that was why we initially discarded that option, but if it there's no other way for this to land in 10.0, we'll continue that way.

**Comment by pedrobaeza** [comment]
@KangOl I have resubmitted the patch in #19153 with a `sudo()` limited to the taxes for avoiding the permission access error. Can you please check?

**Pull request closed without merging**